### PR TITLE
Change Brace from Tex to SVG (#1258)

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Changelog / Overview
Braces are no longer generated by typesetting them with Tex. Now they are created directly as SVGPathMobject, giving them more flexibility.

## Motivation
(see #1258)

## Explanation for Changes

Braces are now created from a SVG path string in exactly the right width. Appart from makin the `Write` animation prettier for braces, this removes the need for parameters `max_num_quads`, `min_num_quads` and `width_multiplier`. To allow adjusting how much the brace is stretched, a new parameter `sharpness` is added. You can see all of this in action here:

```
class BraceScene(MovingCameraScene):
    def construct(self):
        #Zoom in
        self.camera.frame.set(width=5)

        rect = Rectangle(height=1).shift(0.2*UP)
        self.add(rect)
        brace = Brace(rect)
        self.play(Write(brace, run_time=3))

        #Show sharpness values
        tracker = VGroup(Tex("Sharpness: "), DecimalNumber(2)).arrange().scale(0.5)
        brace.put_at_tip(tracker)
        self.play(Write(tracker))
        def update_sharpness(mob, alpha):
            sharpness = 2-alpha*1.5
            mob.become(Brace(rect, sharpness=sharpness))
            tracker[1].set_value(sharpness)
        self.play(UpdateFromAlphaFunc(brace, update_sharpness, run_time=3, rate_func=there_and_back))
        self.play(FadeOut(tracker))

        #Rescale rectangle
        brace.add_updater(lambda mob: mob.become(Brace(rect)))
        self.play(rect.animate.stretch_to_fit_width(0.4))
        brace.clear_updaters()

        #Brace using direction 
        self.play(
            rect.animate.rotate(PI/4),
            FadeOut(brace)
        )
        brace = Brace(rect, direction=DL)
        self.play(Write(brace))

        #get_tex method example
        self.play(Write(brace.get_tex("w")))

        self.wait()
```

https://user-images.githubusercontent.com/26899845/113635813-975f2d00-9671-11eb-9509-c64548d54b35.mp4

## Testing Status
no unit tests but I rendered the test scene using my implementation.

## Further Comments
none

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by MrMallIronmaker** [comment]
I like this change! I follow the motivation you've outlined in #1258.

Can you share the code you've used to make the animation above?

### Expected Behavior:

I like that the sharpness doesn't change when the object scales. 

Are the defaults similar to previous versions of Brace, i.e, the visual sharpness of a brace isn't wildly different before / after this change?

Also, to confirm, larger `sharpness` means smaller curved parts and larger flat parts, right? What's the expected range of values, and is there a nice mapping between the value and the visual result? How easily can I go from a vision or sketch to the sharpness parameter? (This will be important to include in the documentation.)

### Testing:

I don't know why all the tests are failing; it looks like a CI or Poetry issue. If they have the same issue as before (some Poetry can't-find-a-package complaint) then it's safe to ignore.

Have you run the tests locally? (https://docs.manim.community/en/latest/contributing/testing.html)

I think you'll probably have to create a new file for these tests, something like tests/test_graphical_units/test_brace.py

In addition to a smoke test, I think two edge cases should also be included as tests: 
- How are small sharpness values handled in transformations, e.g., if the sharpness parameter remained low while the rectangle shrinks down?
- Does this work when rotated, e.g., along a diagonal line?

Let me know if you need any help, especially with the tests. Looking forward to this change!

**Comment by AntonBallmaier** [comment]
_"Are the defaults similar to previous versions of Brace, i.e, the visual sharpness of a brace isn't wildly different before / after this change?"_
Yes they are. Visually indistinguishable in usual scenarios.

File: manim/mobject/svg/brace.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 __all__ = ["Brace", "BraceLabel", "BraceText", "BraceBetweenPoints"]
 
 
+from manim.utils.color import BLACK
```
Comment (kilacoda) [inline manim/mobject/svg/brace.py:6]: Make this a relative import, and club together with other library imports (i.e. L9-L17)

Comment (AntonBallmaier) [reply manim/mobject/svg/brace.py:6]: Can you explain this to me? I don't know what you mean, sorry.

Comment (AntonBallmaier) [reply manim/mobject/svg/brace.py]: Ok, got it :D

**Comment by AntonBallmaier** [comment]
> Also, to confirm, larger `sharpness` means smaller curved parts and larger flat parts, right?

Right

> What's the expected range of values, and is there a nice mapping between the value and the visual result? How easily can I go from a vision or sketch to the sharpness parameter? (This will be important to include in the documentation.)

Sharpness is simply the factor by which the default LaTeX brace is squished.

> How are small sharpness values handled in transformations, e.g., if the sharpness parameter remained low while the rectangle shrinks down?

If the brace would be to wide using the smallest brace possible, the sharpness is simply increased to get the result. You can see an example of this in the video.

> Does this work when rotated, e.g., along a diagonal line?

Yup. See above.

**Comment by kolibril13** [comment]
Thanks for this pr.
The position of the text is changing a bit in the examples:
https://manimce--1260.org.readthedocs.build/en/1260/examples.html#braceannotation
https://docs.manim.community/en/stable/examples.html#basic-concepts
It does not bother a lot, but I wanted to mention it.
And recently we had the idea that Mobjects should ideally be created counterclockwise (https://github.com/ManimCommunity/manim/pull/1249) , so I'd just like to ask: is the path here counterclockwise as well?

**Comment by AntonBallmaier** [comment]
> The position of the text is changing a bit in the examples:
> https://manimce--1260.org.readthedocs.build/en/1260/examples.html#braceannotation
> https://docs.manim.community/en/stable/examples.html#basic-concepts
> It does not bother a lot, but I wanted to mention it.

Yeah, I noticed that as well and don't exactly know what is causing it. However, I think the way it behaves now is actually closer to what I would expect.

> And recently we had the idea that Mobjects should ideally be created counterclockwise (#1249) , so I'd just like to ask: is the path here counterclockwise as well?

It's counterclockwise. You can see that in the video above (that I forgot to link earlier...)

**Comment by MrMallIronmaker** [comment]
Excellent PR. Looks good to me as soon as tests are added.

**Comment by kolibril13** [comment]
Really great. I think it only needs a test and then it can be merged.

**Comment by AntonBallmaier** [comment]
I would be really grateful if you could help me with the tests, since I have no experience with testing in python.

**Comment by PhotonSpheres** [comment]
_Copy from my answer from Discord:_

I guess you should add a Graphical Unit Test to your SVG Brace, correct? Here is a guide on how to do that in manim: https://docs.manim.community/en/latest/contributing/testing.html

File: manim/mobject/svg/brace.py [hunk 2]
```diff
@@ -104,10 +114,7 @@ def get_tex(self, *tex, **kwargs):
         return tex_mob
 
     def get_tip(self):
-        # Very specific to the LaTeX representation
-        # of a brace, but it's the only way I can think
-        # of to get the tip regardless of orientation.
-        return self.get_all_points()[self.tip_point_index]
+        return self.points[7 * 4]
```
Comment (jsonvillanueva) [inline manim/mobject/svg/brace.py:117]:
```suggestion
        return self.points[28]
```
Also, I don't like that this is hardcoded, but it'd be good to write a comment on what this is doing, or a docstring.

File: manim/mobject/svg/brace.py [hunk 3]
```diff
@@ -51,35 +53,43 @@ def __init__(
...
+        background_stroke_color=BLACK,
         **kwargs
     ):
-        self.width_multiplier = width_multiplier
-        self.max_num_quads = max_num_quads
-        self.min_num_quads = min_num_quads
+        path_string_template = "m0.01216 0c-0.01152 0-0.01216 6.103e-4 -0.01216 0.01311v0.007762c0.06776 0.122 0.1799 0.1455 0.2307 0.1455h{0}c0.03046 3.899e-4 0.07964 0.00449 0.1246 0.02636 0.0537 0.02695 0.07418 0.05816 0.08648 0.07769 0.001562 0.002538 0.004539 0.002563 0.01098 0.002563 0.006444-2e-8 0.009421-2.47e-5 0.01098-0.002563 0.0123-0.01953 0.03278-0.05074 0.08648-0.07769 0.04491-0.02187 0.09409-0.02597 0.1246-0.02636h{0}c0.05077 0 0.1629-0.02346 0.2307-0.1455v-0.007762c-1.78e-6 -0.0125-6.365e-4 -0.01311-0.01216-0.01311-0.006444-3.919e-8 -0.009348 2.448e-5 -0.01091 0.002563-0.0123 0.01953-0.03278 0.05074-0.08648 0.07769-0.04491 0.02187-0.09416 0.02597-0.1246 0.02636h{1}c-0.04786 0-0.1502 0.02094-0.2185 0.1256-0.06833-0.1046-0.1706-0.1256-0.2185-0.1256h{1}c-0.03046-3.899e-4 -0.07972-0.004491-0.1246-0.02636-0.0537-0.02695-0.07418-0.05816-0.08648-0.07769-0.001562-0.002538-0.004467-0.002563-0.01091-0.002563z"
+        default_min_width = 0.90552
```
Comment (jsonvillanueva) [inline manim/mobject/svg/brace.py:64]: Why is this `.90552` and not an integer like 1?

Comment (AntonBallmaier) [reply manim/mobject/svg/brace.py:64]: I scaled the SVG path string so that it is defaults to the size of the previous implementation using Tex.

**Review by MrMallIronmaker — APPROVED** [review]
Tests are good. Thanks!
Now it's just a final merge and a black/flake check.

**Review by behackl — APPROVED** [review]
I like this improvement quite a lot, thank you very much for your contribution.

Have you thought about including the example from the PR description in the docstring of the Brace class? I think it would be a nice example.

**Comment by behackl** [comment]
Including your example would still be nice, but this can also be done in a follow-up PR.

Thanks again for your contribution!

**Comment by AntonBallmaier** [comment]
I add them in a follow-up PR :)
